openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,323 @@
+			if _, ok := c.Resources.Requests[coreapi.ResourceCPU]; !ok {
+				continue
+			}
+
+			cpuRequest := c.Resources.Requests[coreapi.ResourceCPU]
+			cpuRequestInMilli := cpuRequest.MilliValue()
+
+			cpuShares := cm.MilliCPUToShares(cpuRequestInMilli)
+			podAnnotations[cpusAnnotationKey] = fmt.Sprintf("%d", cpuShares)
+			delete(c.Resources.Requests, coreapi.ResourceCPU)
+
+			if c.Resources.Limits == nil {
+				c.Resources.Limits = coreapi.ResourceList{}
+			}
+
+			// multiple the CPU request by 1000, to make sure that the resource will pass integer validation
+			newCPURequest := resource.NewMilliQuantity(cpuRequestInMilli*1000, cpuRequest.Format)
+			c.Resources.Requests[managedCapacityLabel] = *newCPURequest
+			c.Resources.Limits[managedCapacityLabel] = *newCPURequest
+		}
+	}
+}
+
+func isGuaranteed(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request, requestExist := c.Resources.Requests[r]
+
+			if limit.IsZero() {
+				return false
+			}
+
+			if !requestExist {
+				continue
+			}
+
+			// it some corner case, when you set CPU request to 0 the k8s will change it to the value
+			// specified under the limit
+			if r == coreapi.ResourceCPU && request.IsZero() {
+				continue
+			}
+
+			if !limit.Equal(request) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func isBestEffort(containers []coreapi.Container) bool {
+	for _, c := range containers {
+		// only memory and CPU resources are relevant to decide pod QoS class
+		for _, r := range []coreapi.ResourceName{coreapi.ResourceMemory, coreapi.ResourceCPU} {
+			limit := c.Resources.Limits[r]
+			request := c.Resources.Requests[r]
+
+			if !limit.IsZero() || !request.IsZero() {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func getPodQoSClass(containers []coreapi.Container) coreapi.PodQOSClass {
+	if isGuaranteed(containers) {
+		return coreapi.PodQOSGuaranteed
+	}
+
+	if isBestEffort(containers) {
+		return coreapi.PodQOSBestEffort
+	}
+
+	return coreapi.PodQOSBurstable
+}
+
+func isManagementResourceAvailable(nodes []*corev1.Node) bool {
+	for _, node := range nodes {
+		if v, ok := node.Status.Allocatable[managedCapacityLabel]; !ok || v.IsZero() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isNamespaceAllowed(annotations map[string]string) bool {
+	if annotations == nil {
+		return false
+	}
+
+	v, found := annotations[namespaceAllowedAnnotation]
+	if !found {
+		return false
+	}
+
+	for _, workloadType := range strings.Split(v, ",") {
+		if workloadType == workloadTypeManagement {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isPodManaged(annotations map[string]string) bool {
+	if annotations == nil {
+		return false
+	}
+
+	managementAnnotation := fmt.Sprintf("%s/%s", podWorkloadAnnotation, workloadTypeManagement)
```

Comment:
"management" is a special case now, but eventually we want to support this feature generically for other user-defined types of workloads. Is it worth coding this function to be future-proof now?

Reply:
I can pass the workload type as an argument.

Reply:
Passing the type as an argument still means something has to know what types are supported. Could the function look for any annotation with the right prefix and return what it finds so the caller knows how the workload is supposed to be handled? A nil response could mean partitioning is not configured for the pod.